plugin/validator/validator.go

```diff
@@ -0,0 +1,230 @@
+	descriptor "github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"github.com/gogo/protobuf/protoc-gen-gogo/generator"
+)
+
+type plugin struct {
+	*generator.Generator
+	generator.PluginImports
+	regexPkg generator.Single
+	fmtPkg   generator.Single
+}
+
+func NewPlugin() *plugin {
+	return &plugin{}
+}
+
+func (p *plugin) Name() string {
+	return "validator"
+}
+
+func (p *plugin) Init(g *generator.Generator) {
+	p.Generator = g
+}
+
+func (p *plugin) Generate(file *generator.FileDescriptor) {
+	p.PluginImports = generator.NewPluginImports(p.Generator)
+	p.regexPkg = p.NewImport("regexp")
+	p.fmtPkg = p.NewImport("fmt")
+
+	for _, msg := range file.Messages() {
+		if msg.DescriptorProto.GetOptions().GetMapEntry() {
+			continue
+		}
+		if gogoproto.HasValidator(file.FileDescriptorProto, msg.DescriptorProto) {
+			p.generateRegexVars(file, msg)
+			p.generateMessage(file, msg)
+		}
+	}
+}
+
+func (p *plugin) isSupportedInt(field *descriptor.FieldDescriptorProto) bool {
+	switch *(field.Type) {
+	case descriptor.FieldDescriptorProto_TYPE_INT32, descriptor.FieldDescriptorProto_TYPE_INT64:
+		return true
+	case descriptor.FieldDescriptorProto_TYPE_UINT32, descriptor.FieldDescriptorProto_TYPE_UINT64:
+		return true
+	case descriptor.FieldDescriptorProto_TYPE_SINT32, descriptor.FieldDescriptorProto_TYPE_SINT64:
+		return true
+	}
+	return false
+}
+
+func (p *plugin) generateRegexVars(file *generator.FileDescriptor, message *generator.Descriptor) {
+	ccTypeName := generator.CamelCaseSlice(message.TypeName())
+	for _, field := range message.Field {
+		validator := gogoproto.GetValidatorMessage(field)
+		if validator != nil && validator.Regex != nil {
+			fieldName := p.GetFieldName(message, field)
+			p.P(`var `, p.regexName(ccTypeName, fieldName), ` = `, p.regexPkg.Use(), `.MustCompile("`, validator.Regex, `")`)
+		}
+	}
+}
+
+func (p *plugin) generateMessage(file *generator.FileDescriptor, message *generator.Descriptor) {
+	//	proto3 := gogoproto.IsProto3(file.FileDescriptorProto)
+	ccTypeName := generator.CamelCaseSlice(message.TypeName())
+
+	p.P(`func (this *`, ccTypeName, `) Validator() error {`)
+	p.In()
+	for _, field := range message.Field {
+		validator := gogoproto.GetValidatorMessage(field)
+		if validator == nil && !field.IsMessage() {
+			continue
+		}
+		fieldName := p.GetFieldName(message, field)
+		variableName := "this." + fieldName
+		repeated := field.IsRepeated()
+		nullable := gogoproto.IsNullable(field)
+		if repeated {
+			p.P(`for _, item := range `, variableName, `{`)
+			p.In()
+			variableName = "item"
+		}
+		if field.IsString() {
+			p.generateStringValidator(variableName, ccTypeName, fieldName, validator)
+		} else if p.isSupportedInt(field) {
+			p.generateIntValidator(variableName, ccTypeName, fieldName, validator)
+		} else if field.IsMessage() {
+			if nullable {
+				p.P(`if `, variableName, ` != nil {`)
+				p.In()
+			}
+			// TODO(michal): Make the validator test on submessage less expensive.. somehow?
```

Comment:
But this will mean that no one could ever implement there own manual validation method, which is a little inflexible.
I don't think this will have such a big effect on speed.
Maybe when its complete, lets come back to this and do some benchmarks and see if it makes a big difference.